frontend/src/components/viewers/MetricsVisualizations.tsx

```diff
@@ -344,14 +362,227 @@ type ConfidenceMetric = {
+          fullArtifactPath.run,
+          selectedIdColorMap[id],
+        ] as any,
+      };
+      rows.push(row);
+    }
+  }
+  return {
+    columns,
+    rows,
+    selectedConfidenceMetrics: confidenceMetricsDataList,
+  };
+};
+
+const updateRocCurveSelection = (
+  filter: ConfidenceMetricsFilter,
+  maxSelectedRocCurves: number,
+  newSelectedIds: string[],
+): void => {
+  const {
+    selectedIds: oldSelectedIds,
+    setSelectedIds,
+    selectedIdColorMap,
+    setSelectedIdColorMap,
+  } = filter;
+
+  // Convert arrays to sets for quick lookup.
+  const newSelectedIdsSet = new Set(newSelectedIds);
+  const oldSelectedIdsSet = new Set(oldSelectedIds);
+
+  // Find the symmetric difference and intersection of new and old IDs.
+  const addedIds = newSelectedIds.filter(selectedId => !oldSelectedIdsSet.has(selectedId));
+  const removedIds = oldSelectedIds.filter(selectedId => !newSelectedIdsSet.has(selectedId));
+  const sharedIds = oldSelectedIds.filter(selectedId => newSelectedIdsSet.has(selectedId));
+
+  // Restrict the number of selected ROC Curves to a maximum of 10.
+  const numElementsRemaining = maxSelectedRocCurves - sharedIds.length;
+  const limitedAddedIds = addedIds.slice(0, numElementsRemaining);
+  setSelectedIds(sharedIds.concat(limitedAddedIds));
+
+  // Update the color stack and mapping to match the new selected ROC Curves.
+  removedIds.forEach(removedId => {
+    lineColorsStack.push(selectedIdColorMap[removedId]);
+    delete selectedIdColorMap[removedId];
+  });
+  limitedAddedIds.forEach(addedId => {
+    selectedIdColorMap[addedId] = lineColorsStack.pop()!;
+  });
+  setSelectedIdColorMap(selectedIdColorMap);
+};
+
+function reloadRocCurve(
+  filter: ConfidenceMetricsFilter,
+  linkedArtifacts: LinkedArtifact[],
+  setLinkedArtifactsPage: (linkedArtifactsPage: LinkedArtifact[]) => void,
+  request: ListRequest,
+): Promise<string> {
+  // Filter the linked artifacts by run, execution, and artifact display name.
+  const apiFilter = JSON.parse(
+    decodeURIComponent(request.filter || '{"predicates": []}'),
+  ) as ApiFilter;
+  const predicates = apiFilter.predicates?.filter(
+    p => p.key === 'name' && p.op === PredicateOp.ISSUBSTRING,
+  );
+  const substrings = predicates?.map(p => p.string_value?.toLowerCase() || '') || [];
+  const displayLinkedArtifacts = linkedArtifacts.filter(linkedArtifact => {
+    if (filter) {
+      const fullArtifactPath: FullArtifactPath =
+        filter.fullArtifactPathMap[getRocCurveId(linkedArtifact)];
+      for (const sub of substrings) {
+        const executionArtifactName = `${fullArtifactPath.execution.name} > ${fullArtifactPath.artifact.name}`;
+        if (!executionArtifactName.includes(sub) && !fullArtifactPath.run.name.includes(sub)) {
+          return false;
+        }
+      }
+    }
+    return true;
+  });
+
+  let linkedArtifactsPage: LinkedArtifact[] = displayLinkedArtifacts;
+  let nextPageToken: string = '';
+  if (request.pageSize) {
```

Comment:
Unlike the logic in [`RunList.tsx`](https://github.com/kubeflow/pipelines/blob/d48776fb2d9752ca48af14ea6256791d2b167a71/frontend/src/pages/RunList.tsx#L394-L401), I did not want to re-request the MLMD artifacts on each filter as re-constructing the `RunArtifact` objects would be time-consuming. Therefore, I have the `pageToken` represent an incrementing integer which segments the larger linked artifacts list into sub-lists.

Reply:
Sounds good, I've done both here: https://github.com/kubeflow/pipelines/pull/8073/commits/db716e1be197a9dcc858fc56e62b80c8297ca1e4. To clarify, that restriction is 100 artifacts per run, so up to 1000 for 10 runs? Or am I misunderstanding that restriction?

Also, does this same restriction exist on the `RunDetails.tsx` page? https://github.com/kubeflow/pipelines/blob/2a48096dc3cda7df45f15a7e029adda117064cc3/frontend/src/pages/RunDetailsV2.tsx#L105-L110

I'm happy to add a TODO comment there as well.